### ISSUE TITLE
Unnecessary finalizer registration in DefaultConnectionProvider

### ISSUE THREAD

**Issue by SerCeMan** [jOOQ/jOOQ#12174]
Hi, jOQQ Mainainers!

While profiling one of our services, I noticed the following part of the flame graph where a noticeable, although not huge, amount of CPU was spent. 

<img width="1022" alt="Screen Shot 2021-07-14 at 2 22 22 pm" src="https://user-images.githubusercontent.com/1780970/125560799-4656a374-118d-4d4f-b9ef-fb5fec569cce.png">

It seems like even though the `finalize` variable is `false` in the following code, there is a CPU cost of overloading the method. 
https://github.com/jOOQ/jOOQ/blob/6f7c73301466867cf02e8ea6a7eda11d05ed25bf/jOOQ/src/main/java/org/jooq/impl/DefaultConnectionProvider.java#L91-L97

If I remember correctly, there is a way to use different APIs depending on the JDK version in jOOQ. If that's correct, what do you think of using [Cleaner](https://docs.oracle.com/javase/9/docs/api/java/lang/ref/Cleaner.html) here for JDK9+? If not, I'm wondering if there is a way to restructure the code by creating a separate object that would hold the `finalize` method, e.g. something along the lines of.

```java
this.finalizeRef = finalize ? new FinalizerRef(this) : null;
```

If any of these sounds good to you, I'm happy to raise a PR. 

### Expected behavior

No additional CPU is spent when `finalize` is `false`. 

### Actual behavior

Additional CPU is spent when `finalize` is `false`. 

### Steps to reproduce the problem

- If the problem relates to code generation, please post your code generation configuration
- If the problem relates to upgrades, please check if your RDBMS version is still supported by jOOQ: https://www.jooq.org/download/support-matrix
- A complete set of DDL statements can help re-create the setup you're having
- An MCVE can be helpful to provide a complete reproduction case: https://github.com/jOOQ/jOOQ-mcve

### Versions

- jOOQ: `3.14.12`
- Java: 13.0.4
- Database (include vendor): MySQL 8, RDS

**Comment by lukaseder** [comment]
Thanks a lot for your report. Ah yes, I remember having read some blogs (by Aleksey Shipilëv probably?) on the subject. Any non-trivial (i.e. byte code bearing) `finalize()` override can cause quite the overhead in GCs.

I don't think we need to make this too complex for the case where `DefaultConnectionProvider.finalize == false`. Better just create a new `DefaultConnectionProvider` subclass, which doesn't affect the 99.9% of `DefaultConnectionProvider` instances.

**Comment by lukaseder** [comment]
On the other hand, `Object.finalize()` has been deprecated in Java 9. It might be better to just delete the logic entirely. Users should use the `DefaultCloseableDSLContext` as an `AutoCloseable` resource:

```java
try (CloseableDSLContext ctx = DSL.using(url, username, password)) {
  ...
}
```

I wouldn't want to eagerly remove this logic, but it's good to keep this in mind for the future: https://github.com/jOOQ/jOOQ/issues/12178

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.16.0 and 3.15.1 (https://github.com/jOOQ/jOOQ/issues/12179). Thanks again for your report!

**Comment by SerCeMan** [comment]
Thank you @lukaseder for such a speedy change!
